### PR TITLE
Spike and slab2

### PR DESCRIPTION
**Pull request by mschauer**

### CODE REVIEW

File: src/ss_fact.jl [hunk 1]
```diff
@@ -1,13 +1,52 @@
 using SparseArrays
 const SA = SparseArrays
 
+"""
+    τ = freezing_time(x, θ)
+computes the hitting time of a 1d particle with
```
Comment (mschauer) [inline src/ss_fact.jl:6]:
```suggestion

computes the hitting time of a 1d particle with
```

File: src/ss_fact.jl [hunk 2]
```diff
@@ -1,13 +1,52 @@
...
+    end
+end
+
 
-function smove_forward!(i::Int, t, x, θ, t′, Z::ZigZag)
-    t[i], x[i], θ = t′, x[i] + θ[i]*(t′ - t[i]), θ
+
+"""
+    t, x, θ = ssmove_forward!(t, x, θ, t′, Z::Union{BouncyParticle, ZigZag})
+moves forward only the non_frozen particles
```
Comment (mschauer) [inline src/ss_fact.jl:21]:
```suggestion

moves forward only the non_frozen particles
```
leave a line

File: src/ss_fact.jl [hunk 3]
```diff
@@ -42,32 +87,35 @@ function sspdmp_inner!(Ξ, G, G2, ∇ϕ, t, x, θ, Q, c, b, t_old, f, θf, (acc,
             t_old[i] = t[i]
             f[i] = false
             Q[i] = t[i] - log(rand())/κ
```
Comment (mschauer) [inline src/ss_fact.jl:89]: We need to make kappa a vector
```suggestion
            Q[i] = t[i] - log(rand())/κ[i]
```

Comment (SebaGraz) [reply src/ss_fact.jl:91]: :+1:

File: src/ss_fact.jl [hunk 4]
```diff
@@ -76,9 +124,9 @@ function sspdmp_inner!(Ξ, G, G2, ∇ϕ, t, x, θ, Q, c, b, t_old, f, θf, (acc,
                     acc = num = 0
                     adapt!(c, i, factor)
                 end
-                t, x, θ = smove_forward!(G, i, t, x, θ, t′, F) # neighbours
-                t, x, θ = smove_forward!(G2, i, t, x, θ, t′, F) # neighbours of neightbours \ neighbours
-                θ = reflect!(i, x, θ, F)
+                # already done above t, x, θ = smove_forward!(G, i, t, x, θ, t′, F) # neighbours
+                t, x, θ = ssmove_forward!(G2, i, t, x, θ, t′, F) # neighbours of neightbours \ neighbours
+                θ[i] = -θ[i]
```
Comment (mschauer) [inline src/ss_fact.jl:131]:
```
θ = reflect!(i, x, θ, F)
```
in case `θ` is a vector of vectors

File: src/ss_fact.jl [hunk 5]
```diff
@@ -87,10 +135,9 @@ function sspdmp_inner!(Ξ, G, G2, ∇ϕ, t, x, θ, Q, c, b, t_old, f, θf, (acc,
                     end
                 end
             else # was an event time from upperbound -> nothing happens
-                t, x, θ = smove_forward!(G, i, t, x, θ, t′, F) # neighbours
```
Comment (mschauer) [inline src/ss_fact.jl:90]: Not needed?

File: src/ss_fact.jl [hunk 6]
```diff
@@ -42,32 +87,35 @@ function sspdmp_inner!(Ξ, G, G2, ∇ϕ, t, x, θ, Q, c, b, t_old, f, θf, (acc,
...
                     t_old[j] = t[j]
-                    queue_time!(Q, t, x, θ, j, b, f, F)
+                    Q = queue_time!(Q, t, x, θ, j, b, f, F)
                 end
             end
         else # was either a reflection time or an event time from the upper bound
-            t, x, θ = smove_forward!(G, i, t, x, θ, t′, F) #TO CHANGE
-            l = sλ(∇ϕ, i, t, x, θ, t′, F, args...)
-            l, lb = sλ(∇ϕ, i, t, x, θ, t′, F, args...), sλ̄(b[i], t[i] - t_old[i])
+            t, x, θ = smove_forward!(G, i, t, x, θ, t′, F) # neighbours
```
Comment (mschauer) [inline src/ss_fact.jl:117]: ssmove?

File: src/ss_fact.jl [hunk 7]
```diff
@@ -18,56 +62,66 @@ function queue_time!(Q, t, x, θ, i, b, f, Z::ZigZag)
...
+Inner loop of the sticky ZigZag sampler. `G[i]` is the set of indices used to derive the
+bounding rate λbar_i and `G2` are the indices k in A_j for all j : i in Aj (neighbours of neighbours)
 
-
+If ∇ϕ is not self moving, then it is assumed that ∇ϕ[x, i] is function of x_i
+with i in G[i].
+"""
 function sspdmp_inner!(Ξ, G, G2, ∇ϕ, t, x, θ, Q, c, b, t_old, f, θf, (acc, num),
-        F::ZigZag, κ, args...; strong_upperbounds = false, factor=1.5, adapt=false)
+        F::ZigZag, κ::Vector{Float64}, args...; strong_upperbounds = false, factor=1.5, adapt=false)
```
Comment (mschauer) [inline src/ss_fact.jl:79]: Don't restrict the arguments, could also be a Sparse Vector etc.

File: src/ss_fact.jl [hunk 8]
```diff
@@ -1,13 +1,57 @@
...
+function ssmove_forward!(t, x, θ, t′, Z::Union{BouncyParticle, ZigZag})
+    for i in eachindex(x)
+        if θ[i] != 0.0
+            t[i], x[i] = t′, x[i] + θ[i]*(t′ - t[i])
+        end
+    end
     t, x, θ
 end
+"""
+    t, x, θ = smove_forward!(G, i, t, x, θ, t′, Z::Union{BouncyParticle, ZigZag})
```
Comment (mschauer) [inline src/ss_fact.jl:34]:
```suggestion
    t, x, θ = ssmove_forward!(G, i, t, x, θ, t′, Z::Union{BouncyParticle, ZigZag})
    
```

File: src/ss_fact.jl (hunk 7 above)
Comment (mschauer) [inline src/ss_fact.jl:79]:
```suggestion
        F::ZigZag, κ, args...; strong_upperbounds = false, factor=1.5, adapt=false)
```

File: src/ss_fact.jl [hunk 9]
```diff
@@ -18,56 +62,66 @@ function queue_time!(Q, t, x, θ, i, b, f, Z::ZigZag)
...
     while true
         ii, t′ = peek(Q)
         refresh = ii > n
         i = ii - refresh*n
         refresh && error("refreshment not implemented")
-        #t, x, θ = smove_forward!(G, i, t, x, θ, t′, F)
-        t, x, θ = smove_forward!(i, t, x, θ, t′, F) # move only coordinate i
-        if f[i] # to be frozen
+        if f[i] # case 1) to be frozen
+            t, x, θ = smove_forward!(i, t, x, θ, t′, F) # move only coordinate i
```
Comment (mschauer) [inline src/ss_fact.jl:87]:
```suggestion
            t, x, θ = ssmove_forward!(i, t, x, θ, t′, F) # move only coordinate i
```

File: src/ss_fact.jl [hunk 10]
```diff
@@ -18,56 +62,66 @@ function queue_time!(Q, t, x, θ, i, b, f, Z::ZigZag)
...
                     t_old[j] = t[j]
-                    queue_time!(Q, t, x, θ, j, b, f, F)
+                    Q = queue_time!(Q, t, x, θ, j, b, f, F)
                 end
             end
         else # was either a reflection time or an event time from the upper bound
-            t, x, θ = smove_forward!(G, i, t, x, θ, t′, F) #TO CHANGE
-            l = sλ(∇ϕ, i, t, x, θ, t′, F, args...)
-            l, lb = sλ(∇ϕ, i, t, x, θ, t′, F, args...), sλ̄(b[i], t[i] - t_old[i])
+            t, x, θ = smove_forward!(G, i, t, x, θ, t′, F) # neighbours
```
Comment (mschauer) [inline src/ss_fact.jl:121]:
```suggestion
            t, x, θ = ssmove_forward!(G, i, t, x, θ, t′, F) # neighbours
```

File: src/sfact.jl [hunk 11]
```diff
@@ -86,7 +87,7 @@ function spdmp_inner!(Ξ, G, G2, ∇ϕ, t, x, θ, Q, c, b, t_old, (acc, num),
                     adapt!(c, i, factor)
                 end
                 t, x, θ = smove_forward!(G2, i, t, x, θ, t′, F)
-                θ = reflect!(i, ∇ϕi, x, θ, F)
+                θ = reflect!(i, x, θ, F)
```
Comment (mschauer) [inline src/sfact.jl:90]:
```suggestion
                θ = reflect!(i, ∇ϕi, x, θ, F)
```

Comment (mschauer) [reply src/sfact.jl:90]: For BPS within Zig-Zag

File: src/ss_fact.jl [hunk 12]
```diff
@@ -1,13 +1,58 @@
...
 
+"""
+    τ = freezing_time(x, θ)
 
-function smove_forward!(i::Int, t, x, θ, t′, Z::ZigZag)
-    t[i], x[i], θ = t′, x[i] + θ[i]*(t′ - t[i]), θ
+computes the hitting time of a 1d particle with
+constant velocity `θ` to hit 0 given the position `x`
+"""
+function freezing_time(x, θ)
```
Comment (mschauer) [inline src/ss_fact.jl:10]: Shouldn't this dispatch on `ZigZag`?

File: src/ss_fact.jl (hunk 9 above)
Comment (mschauer) [reply src/ss_fact.jl:88]: Check?

File: src/ss_fact.jl [hunk 13]
```diff
@@ -76,8 +131,8 @@ function sspdmp_inner!(Ξ, G, G2, ∇ϕ, t, x, θ, Q, c, b, t_old, f, θf, (acc,
                     acc = num = 0
                     adapt!(c, i, factor)
                 end
-                t, x, θ = smove_forward!(G, i, t, x, θ, t′, F) # neighbours
-                t, x, θ = smove_forward!(G2, i, t, x, θ, t′, F) # neighbours of neightbours \ neighbours
+                # already done above t, x, θ = smove_forward!(G, i, t, x, θ, t′, F) # neighbours
+                t, x, θ = ssmove_forward!(G2, i, t, x, θ, t′, F) # neighbours of neightbours \ neighbours
                 θ = reflect!(i, x, θ, F)
```
Comment (mschauer) [inline src/ss_fact.jl:136]:
```suggestion
                θ = reflect!(i, ∇ϕi, x, θ, F)
```
check if `∇ϕi` is still the gradient in `x`, `i`

**Pull request merged**